processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -0,0 +1,1321 @@
+    final long lastTimestamp = index.getMaxTime().getMillis();
+    if (!(dataInterval.contains(firstTimestamp) && dataInterval.contains(lastTimestamp))) {
+      throw new IAE(
+          "interval[%s] does not encapsulate the full range of timestamps[%s, %s]",
+          dataInterval,
+          new DateTime(firstTimestamp),
+          new DateTime(lastTimestamp)
+      );
+    }
+
+    if (!outDir.exists()) {
+      outDir.mkdirs();
+    }
+    if (!outDir.isDirectory()) {
+      throw new ISE("Can only persist to directories, [%s] wasn't a directory", outDir);
+    }
+
+    log.info("Starting persist for interval[%s], rows[%,d]", dataInterval, index.size());
+    return merge(
+        Arrays.<IndexableAdapter>asList(
+            new IncrementalIndexAdapter(
+                dataInterval,
+                index,
+                indexSpec.getBitmapSerdeFactory().getBitmapFactory()
+            )
+        ),
+        index.getMetricAggs(),
+        outDir,
+        segmentMetadata,
+        indexSpec,
+        progress
+    );
+  }
+
+  public File convert(final File inDir, final File outDir, final IndexSpec indexSpec) throws IOException
+  {
+    return convert(inDir, outDir, indexSpec, new LoggingProgressIndicator("index_convert"));
+  }
+
+  public File convert(
+      final File inDir, final File outDir, final IndexSpec indexSpec, final ProgressIndicator progress
+  ) throws IOException
+  {
+    try (QueryableIndex index = indexIO.loadIndex(inDir)) {
+      final IndexableAdapter adapter = new QueryableIndexIndexableAdapter(index);
+      return makeIndexFiles(
+          ImmutableList.of(adapter),
+          outDir,
+          progress,
+          Lists.newArrayList(adapter.getDimensionNames()),
+          Lists.newArrayList(adapter.getMetricNames()),
+          null,
+          new Function<ArrayList<Iterable<Rowboat>>, Iterable<Rowboat>>()
+          {
+            @Nullable
+            @Override
+            public Iterable<Rowboat> apply(ArrayList<Iterable<Rowboat>> input)
+            {
+              return input.get(0);
+            }
+          },
+          indexSpec
+      );
+    }
+  }
+
+  public File mergeQueryableIndex(
+      List<QueryableIndex> indexes, final AggregatorFactory[] metricAggs, File outDir, IndexSpec indexSpec
+  ) throws IOException
+  {
+    return mergeQueryableIndex(indexes, metricAggs, outDir, indexSpec, new LoggingProgressIndicator("index_merge"));
+  }
+
+  public File mergeQueryableIndex(
+      List<QueryableIndex> indexes,
+      final AggregatorFactory[] metricAggs,
+      File outDir,
+      IndexSpec indexSpec,
+      ProgressIndicator progress
+  ) throws IOException
+  {
+    // We are materializing the list for performance reasons. Lists.transform
+    // only creates a "view" of the original list, meaning the function gets
+    // applied every time you access an element.
+    List<IndexableAdapter> indexAdapteres = Lists.newArrayList(
```

Comment:
how much of a performance difference does this actually make? I think things are fine both ways though.